src/irmin-pack/pack_store.ml

```diff
@@ -136,83 +165,118 @@ module Maker
 
-    exception Invalid_read
+    let _check_key k v = check_hash (Key.to_hash k) v
 
     let io_read_and_decode ~off ~len t =
-      if (not (IO.readonly t.pack.block)) && off > IO.offset t.pack.block then
-        raise Invalid_read;
       let buf = Bytes.create len in
       let n = IO.read t.pack.block ~off buf in
-      if n <> len then raise Invalid_read;
-      let hash off = io_read_and_decode_hash ~off t in
+      if n <> len then
+        invalid_read "Read %d bytes (at offset %a) but expected %d" n Int63.pp
+          off len;
+      let key_of_offset off =
+        let hash = io_read_and_decode_hash ~off t in
+        Pack_key.v ~hash ~offset:off ~length:len
+      in
       let dict = Dict.find t.pack.dict in
-      Val.decode_bin ~hash ~dict (Bytes.unsafe_to_string buf) 0
+      Val.decode_bin ~key_of_offset ~dict (Bytes.unsafe_to_string buf) 0
 
     let pp_io ppf t =
       let name = Filename.basename (Filename.dirname (IO.name t.pack.block)) in
       let mode = if t.readonly then ":RO" else "" in
       Fmt.pf ppf "%s%s" name mode
 
-    let unsafe_find ~check_integrity t k =
-      Log.debug (fun l -> l "[pack:%a] find %a" pp_io t pp_hash k);
+    let unsafe_find ~check_integrity:_ t k =
+      Log.debug (fun l -> l "[pack:%a] find %a" pp_io t pp_key k);
       Stats.incr_finds ();
-      match Tbl.find t.staging k with
+      let hash = Key.to_hash k in
+      match Tbl.find t.staging hash with
       | v ->
-          Lru.add t.lru k v;
+          Lru.add t.lru hash v;
           Some v
       | exception Not_found -> (
-          match Lru.find t.lru k with
+          match Lru.find t.lru hash with
           | v -> Some v
           | exception Not_found -> (
               Stats.incr_cache_misses ();
-              match Index.find t.pack.index k with
-              | None -> None
-              | Some (off, len, _) ->
-                  let v = snd (io_read_and_decode ~off ~len t) in
-                  (if check_integrity then
-                   check_key k v |> function
-                   | Ok () -> ()
-                   | Error (expected, got) ->
-                       Fmt.failwith "corrupted value: got %a, expecting %a."
-                         pp_hash got pp_hash expected);
-                  Lru.add t.lru k v;
-                  Some v))
+              (* XXX *)
+              (* match Index.find t.pack.index k with
+               * | None -> None
+               * | Some (off, len, _) -> *)
+              let off, len = Key.(to_offset k, to_length k) in
+              if off >= IO.offset t.pack.block then
+                (* This key is from a different store instance, referencing an
+                   offset that is not yet visible to this one. It's possible
+                   that the key _is_ a valid pointer into the same store, but
+                   the offset just hasn't been flushed to disk yet, so we return
+                   [None]. *)
+                None
```

Comment:
When I first came across this, raising an error failed a test in which a read-only instance received a read-write key and was "waiting" to observe it via sync. At the time, I decided this was OK because keys belong to store instances (which may have multiple handles) rather than to individual handles; perhaps forbidding this would be cleaner though.

> In that case shouldn't this entry be contained in `Tbl`?

`t.staging` and `t.lru` are both memoised using `readonly` as a key, so RO instances don't observe changes made by RW ones in this way. (But perhaps they shoud...)